mlir/lib/Dialect/IRDL/IR/IRDL.cpp

```diff
@@ -74,13 +74,62 @@ static void printSingleBlockRegion(OpAsmPrinter &p, Operation *op,
   if (!region.getBlocks().front().empty())
     p.printRegion(region);
 }
+static llvm::LogicalResult isSnakeCase(llvm::StringRef in, mlir::Operation *loc,
+                                       const Twine &label) {
+  if (in.empty())
+    return loc->emitError("name of ") << label << " is empty";
+
+  bool allowUnderscore = false;
+  for (auto &elem : in) {
+    if (elem == '_') {
+      if (!allowUnderscore)
+        return loc->emitError("name of ")
+               << label << " should not contain leading or double underscores";
+    } else {
+      if (!isalnum(elem))
+        return loc->emitError("name of ")
+               << label
+               << " must contain only lowercase letters, digits and "
+                  "underscores";
+
+      if (llvm::isUpper(elem))
+        return loc->emitError("name of ")
+               << label << " should not contain uppercase letters";
+    }
+
+    allowUnderscore = elem != '_';
+  }
+
+  return success();
+}
 
 LogicalResult DialectOp::verify() {
   if (!Dialect::isValidNamespace(getName()))
     return emitOpError("invalid dialect name");
+  if (failed(isSnakeCase(getSymName(), getOperation(), "dialect")))
+    return failure();
+
   return success();
 }
 
+LogicalResult OperationOp::verify() {
+  return isSnakeCase(getSymName(), getOperation(), "operation");
+}
+
+LogicalResult TypeOp::verify() {
+  auto symName = getSymName();
+  if (symName.front() == '!')
```

Comment:
This is no bueno but we have to due to symbols. Once this PR is merged upstream I'll fix it myself.